src/System.Private.CoreLib/shared/System/Number.Parsing.cs

```diff
@@ -1468,64 +1468,88 @@ internal static decimal ParseDecimal(ReadOnlySpan<char> value, NumberStyles styl
                 // the scale to 0 if the scale was previously positive (previously, such cases were unparsable to a bug.)
-                if (e > 0)
+                value = new decimal(0, 0, 0, sign, (byte)Math.Clamp(-e, 0, 28));
+                return true;
+            }
+
+            if (e > DecimalPrecision)
+                return false;
+
+            ulong low64 = 0;
+            while (e > -28)
+            {
+                e--;
+                low64 *= 10;
+                low64 += c - '0';
+                c = *++p;
+                if (low64 >= ulong.MaxValue / 10)
+                    break;
+                if (c == 0)
                 {
-                    e = 0;
+                    while (e > 0)
+                    {
+                        e--;
+                        low64 *= 10;
+                        if (low64 >= ulong.MaxValue / 10)
+                            break;
+                    }
+                    break;
                 }
             }
-            else
+
+            uint high = 0;
+            while ((e > 0 || c != 0 && e > -28) &&
+              (high < uint.MaxValue / 10 || high == uint.MaxValue / 10 && (low64 < 0x99999999_99999999 || low64 == 0x99999999_99999999 && c <= '5')))
             {
-                if (e > DecimalPrecision)
-                    return false;
+                // multiply by 10
+                ulong tmpLow = (uint)low64 * 10UL;
+                ulong tmp64 = (uint)(low64 >> 32) * 10UL + (tmpLow >> 32);
+                low64 = (uint)tmpLow + (tmp64 << 32);
+                high = (uint)(tmp64 >> 32) + high * 10;
 
-                while (((e > 0) || ((*p != 0) && (e > -28))) &&
-                       ((d.High < 0x19999999) || ((d.High == 0x19999999) &&
-                                                  ((d.Mid < 0x99999999) || ((d.Mid == 0x99999999) &&
-                                                                            ((d.Low < 0x99999999) || ((d.Low == 0x99999999) &&
-                                                                                                      (*p <= '5'))))))))
+                if (c != 0)
                 {
-                    decimal.DecMul10(ref d);
-                    if (*p != 0)
-                        decimal.DecAddInt32(ref d, (uint)(*p++ - '0'));
-                    e--;
+                    c -= '0';
+                    if ((low64 += c) < c)
+                        high++;
+                    c = *++p;
                 }
+                e--;
+            }
 
-                if (*p++ >= '5')
+            if (c >= '5')
+            {
+                if (c == '5' && (low64 & 1) == 0)
                 {
-                    bool round = true;
-                    if ((*(p - 1) == '5') && ((*(p - 2) % 2) == 0))
+                    c = *++p;
+                    // Check if previous digit is even, only if the when we are unsure whether hows to do
```

Comment:
`only if the when we are unsure whether hows to do` does not make sense. (I know it was there before your change as well.)